src/diffusers/models/unet_2d_condition.py

```diff
@@ -310,8 +310,12 @@ def forward(
         timesteps = timestep
         if not torch.is_tensor(timesteps):
             # TODO: this requires sync between CPU and GPU. So try to pass timesteps as tensors if you can
-            timesteps = torch.tensor([timesteps], dtype=torch.long, device=sample.device)
-        elif torch.is_tensor(timesteps) and len(timesteps.shape) == 0:
+            if sample.device.type == "mps":
+                # mps does not support int64 or float64
+                timesteps = torch.tensor([timesteps], dtype=torch.float32, device=sample.device)
+            else:
+                timesteps = torch.tensor([timesteps], dtype=torch.long, device=sample.device)
```

Comment:
Should we always use `float32`?

Reply:
Hmm, I think timesteps is used as a look up later no? Having an if statement for "mps" looks good to me!

Reply:
Might have unforeseen consequences on reproducibility, let's run the slow tests to check

Reply:
Yeah, for example `AdaLayerNorm` uses `nn.Embedding` for timesteps

Reply:
Yeah, probably! We also have a silent corner case here: this step should work when timesteps are `float32` (when they come from K-LMS for instance). They're usually passed as a tensor already, so this was working implicitly before

Reply:
@anton-l so would this be better?

```Python
            is_mps = sample.device.type == "mps"
            if torch.is_floating_point(timesteps):
                dtype = torch.float32 if is_mps else torch.float64
            else:
                dtype = torch.int32 if is_mps else torch.int64
```

Reply:
Perfect :+1: